sweyer_plugin/lib/sweyer_plugin.dart

```diff
@@ -0,0 +1,183 @@
+import 'dart:typed_data';
+import 'dart:ui';
+
+import 'package:equatable/equatable.dart';
+import 'package:sweyer_plugin/sweyer_plugin_method_channel.dart';
+import 'package:uuid/uuid.dart';
+
+import 'sweyer_plugin_platform_interface.dart';
+
+/// A song from the native media store.
+abstract class MediaStoreSong {
+  /// The id of this song in the native media store.
+  int get sourceId;
+
+  /// Absolute filesystem path to the media item on disk.
+  String? get filesystemPath;
+}
+
+/// An album from the native media store.
+abstract class MediaStoreAlbum {}
+
+/// An artist from the native media store.
+abstract class MediaStoreArtist {}
+
+/// A playlist from the native media store.
+abstract class MediaStorePlaylist {
+  /// The id of this playlist in the native media store.
+  int get id;
+
+  /// A list of ids of songs that this playlist contains.
+  List<int> get songIds;
+}
+
+/// A genre from the native media store.
+abstract class MediaStoreGenre {}
+
+abstract class SweyerPluginException with EquatableMixin {
+  const SweyerPluginException();
+  String get value;
+
+  @override
+  List<Object?> get props => [value];
+
+  /// Generic error.
+  static const unexpected = SweyerMethodChannelException.unexpected;
+
+  /// On Android 30 requests like `MediaStore.createDeletionRequest` require
+  /// calling `startIntentSenderForResult`, which might throw this exception.
+  static const intentSender = SweyerMethodChannelException.intentSender;
+
+  static const io = SweyerMethodChannelException.io;
+
+  /// API is unavailable on current SDK level.
+  static const sdk = SweyerMethodChannelException.sdk;
+
+  /// Operation cannot be performed because there's no such playlist.
+  static const playlistNotExists = SweyerMethodChannelException.playlistNotExists;
+}
+
+const _uuid = Uuid();
+
+/// Class to cancel the [ContentChannel.loadAlbumArt].
+class CancellationSignal {
+  CancellationSignal() : _id = _uuid.v4();
+  final String _id;
+
+  /// Cancel loading of an album art.
+  Future<void> cancel() => SweyerPluginPlatform.instance.cancelAlbumArtLoad(id: _id);
+}
+
+abstract class SweyerPlugin {
```

Comment:
Let's not make it abstract with static methods, but rather a regular class and expose it as singletone for now (until we have a proper DI setup in the app).

As it was done before.

Reply:
Hm, this seems a bit unnecessary.
The abstract class with static methods is just an interface to the app and comes from the flutter federated plugin example. Its main job is to delegate the actual work to the `SweyerPluginPlatform` instance, which is where the DI can come in.

But this class is not meant to be instantiated and never has any state that would require an instance. In fact, maybe I should add a private constructor to prevent instantiation? Is there anything I'm overlooking here where the instance would be beneficial?

Reply:
> The abstract class with static methods is just an interface to the app and comes from the flutter federated plugin example.

Yes, but abstract class with static methods is almost always a bad idea, because it doesn't have an interface, therefore it's not mockable, fakeable, testetable and is inherently a signletone (but worse).

In the past I rewrote the entirety of the app from this pattern just to make it testable https://github.com/nt4f04uNd/sweyer/pull/21

Let's not add additional tech-debt.

> But this class is not meant to be instantiated and never has any state that would require an instance.

Why not?

> Is there anything I'm overlooking here where the instance would be beneficial?

In the future everything should be in the DI, so classes don't have implicit dependencies through singletones or patterns like locator.

More of my thoughts on why this is bad is here https://github.com/nt4f04uNd/sweyer/issues/81

Reply:
> Yes, but abstract class with static methods is almost always a bad idea, because it doesn't have an interface, therefore it's not mockable, fakeable, testetable and is inherently a signletone (but worse).

But the class is merely a wrapper around the `SweyerPluginPlatform`, which is the singleton in this case and already uses the pattern you are suggesting. I think that this class is a not singelton, it is merely a collection of static functions and therefore not more than a namespace (which is why it should never be instantiated).

Converting this namespace into a singleton which just delegates to another singleton is not worth it in my opinion. Because this class is just a wrapper, there is no use to mocking it when we can just mock the underlying `SweyerPluginPlatform`.

> In the future everything should be in the DI, so classes don't have implicit dependencies through singletones or patterns like locator.

I agree, we would be able to use `riverpod` in the static functions to resolve the `SweyerPluginPlatform`.

If you still think there is a need to mock the wrapping functions in the future, I'll convert the code to use the singleton pattern here. Let me know.

Reply:
> But the class is merely a wrapper around the `SweyerPluginPlatform`, which is the singleton in this case and already uses the pattern you are suggesting. I think that this class is a not singelton, it is merely a collection of static functions and therefore not more than a namespace (which is why it should never be instantiated).

The user shouldn't know or care about it.

Ideally we shouldn't have been faking the `FakeSweyerPluginPlatform`, but `SweyerPlugin`, but it's fine to leave as-is right now.

> If you still think there is a need to mock the wrapping functions in the future, I'll convert the code to use the singleton pattern here. Let me know.

Yes, I still think this would be valuable.